src/OrchardCore/OrchardCore.DisplayManagement/Razor/RazorPage.cs

```diff
@@ -80,6 +80,35 @@ public IShapeFactory Factory
         /// </summary>
         /// <param name="shape">The shape.</param>
         public Task<IHtmlContent> DisplayAsync(dynamic shape)
+        {
+            EnsureDisplayHelper();
+
+            if (shape is IShape s)
+            {
+                return _displayHelper.ShapeExecuteAsync(s);
+            }
+            else
+            {
+                if (shape is IHtmlContent hc)
+                {
+                    return Task.FromResult(hc);
+                }
+                else if (shape is string str)
+                {
+                    return Task.FromResult<IHtmlContent>(new HtmlString(HtmlEncoder.Encode(str)));
+                }
+                else
+                {
+                    throw new ArgumentException("DisplayAsync requires an instance of IShape");
+                }
+            }
+        }
```

Comment:
At some point i did something like this but reverted it as implemented by `ShapeExecuteAsync()` => `DefaultHtmlDisplay.ExecuteAsync()` where i fixed the use case of `ZoneOnDemand` not considered as null if passed as an `IShape`. But yes, here it is done immediately and returns something if it is a string.

But why not keeping the same logic as in `DefaultHtmlDisplay.CoerceHtmlString()`, e.g. using a `StringHtmlContent` and not throwing if neither an `IHtmlContent` nor a string, here the copy of `CoerceHtmlString()`.

        private static IHtmlContent CoerceHtmlString(object value)
        {
            // So here just need to check if null as ZoneOnDemand is an IShape
            if (value == null || value is ZoneOnDemand)
            {
                return HtmlString.Empty;
            }

            if (value is IHtmlContent result)
            {
                return result;
            }

            // Convert to a string and HTML-encode it <= Remove / Correct the comment saying it will be encoded.
            return new StringHtmlContent(value.ToString());
        }

Reply:
I'd like that we find where it's not correctly used and fix it. Next step for instance would be to use strongly typed models in the forms for instance. And finally remove DisplayAsync(dynamic). Or at least have an "object" overload to handle explicitly "anything", and not "magic"

Reply:
@sebastienros

Okay, what i also suggested here (also in `Page.cs`), regarding what we do in other places, is something like this

        public Task<IHtmlContent> DisplayAsync(dynamic shape)
        {
            if (shape is IShape s)
            {
                EnsureDisplayHelper();
                return _displayHelper.ShapeExecuteAsync(s);
            }

            if (shape == null)
            {
                return Task.FromResult<IHtmlContent>(HtmlString.Empty);
            }

            if (shape is IHtmlContent hc)
            {
                return Task.FromResult(hc);
            }

            // Not necessary as managed below, but maybe faster
            // if (shape is string str)
            // {
            //    return Task.FromResult<IHtmlContent>(new StringHtmlContent(str));
            // }

            return Task.FromResult<IHtmlContent>(new StringHtmlContent(shape.ToString()));
        }

Reply:
At least maybe use `new StringHtmlContent(str)`
in place of `new HtmlString(HtmlEncoder.Encode(str)`